File: src/domain/file/file_service.ts
import { FileInfo, Pagination } from '../../core';
import { getApiClient } from '../../infrastructure';
import { BluetoothService } from '../bluetooth/bluetooth_service';

/**
 * 文件服务
 * 处理文件相关的业务逻辑
 */
export class FileService {
  private static instance: FileService;
  private bluetoothService: BluetoothService;

  private constructor() {
    this.bluetoothService = BluetoothService.getInstance();
  }

  /**
   * 获取文件服务实例
   * @returns 文件服务实例
   */
  static getInstance(): FileService {
    if (!FileService.instance) {
      FileService.instance = new FileService();
    }
    return FileService.instance;
  }

  /**
   * 获取认证头
   * @returns 认证头
   */
  private async getAuthHeader(): Promise<Record<string, string>> {
    const deviceId = await this.bluetoothService.getDeviceId();
    const totp = await this.bluetoothService.getTotp();
    
    return {
      'Id': deviceId,
      'Totp': totp
    };
  }

  /**
   * 列出文件
   * @param path 路径
   * @param pagination 分页参数
   * @returns 文件列表
   */
  async listFiles(path: string = '/', pagination: Pagination = { page: 1, pageSize: 20 }): Promise<FileInfo[]> {
    try {
      const apiClient = getApiClient();
      const authHeader = await this.getAuthHeader();
      
      const files = await apiClient.get<FileInfo[]>('/files', {
        params: {
          path,
          page: pagination.page,
          page_size: pagination.pageSize
        },
        headers: authHeader
      });
      
      return files;
    } catch (error) {
      throw error;
    }
  }

  /**
   * 创建目录
   * @param path 路径
   */
  async createDirectory(path: string): Promise<void> {
    try {
      const apiClient = getApiClient();
      const authHeader = await this.getAuthHeader();
      
      await apiClient.post('/files/directories', { path }, {
        headers: authHeader
      });
    } catch (error) {
      throw error;
    }
  }

  /**
   * 删除文件/目录
   * @param path 路径
   */
  async delete(path: string): Promise<void> {
    try {
      const apiClient = getApiClient();
      const authHeader = await this.getAuthHeader();
      
      await apiClient.delete(`/files/${encodeURIComponent(path)}`, {
        headers: authHeader
      });
    } catch (error) {
      throw error;
    }
  }

  /**
   * 重命名
   * @param oldPath 旧路径
   * @param newPath 新路径
   */
  async rename(oldPath: string, newPath: string): Promise<void> {
    try {
      const apiClient = getApiClient();
      const authHeader = await this.getAuthHeader();
      
      await apiClient.post('/files/rename', {
        old_path: oldPath,
        new_path: newPath
      }, {
        headers: authHeader
      });
    } catch (error) {
      throw error;
    }
  }

  /**
   * 移动
   * @param fromPath 源路径
   * @param toPath 目标路径
   */
  async move(fromPath: string, toPath: string): Promise<void> {
    try {
      const apiClient = getApiClient();
      const authHeader = await this.getAuthHeader();
      
      await apiClient.post('/files/move', {
        from_path: fromPath,
        to_path: toPath
      }, {
        headers: authHeader
      });
    } catch (error) {
      throw error;
    }
  }

  /**
   * 复制
   * @param fromPath 源路径
   * @param toPath 目标路径
   */
  async copy(fromPath: string, toPath: string): Promise<void> {
    try {
      const apiClient = getApiClient();
      const authHeader = await this.getAuthHeader();
      
      await apiClient.post('/files/copy', {
        from_path: fromPath,
        to_path: toPath
      }, {
        headers: authHeader
      });
    } catch (error) {
      throw error;
    }
  }

  /**
   * 下载文件
   * @param path 路径
   * @param onProgress 进度回调
   * @returns 文件数据
   */
  async download(path: string, onProgress?: (progress: number) => void): Promise<Blob> {
    try {
      const apiClient = getApiClient();
      await this.getAuthHeader();
      
      return await apiClient.download(`/files/download/${encodeURIComponent(path)}`, onProgress) as Blob;
    } catch (error) {
      throw error;
    }
  }

  /**
   * 上传文件
   * @param path 路径
   * @param file 文件
   * @param onProgress 进度回调
   */
  async upload(path: string, file: File, onProgress?: (progress: number) => void): Promise<void> {
    try {
      const apiClient = getApiClient();
      await this.getAuthHeader();
      
      const formData = new FormData();
      formData.append('file', file);
      
      await apiClient.upload(`/files/upload/${encodeURIComponent(path)}`, formData, onProgress);
    } catch (error) {
      throw error;
    }
  }
}

File: src-tauri/src/domain/file/file_service.rs
use async_trait::async_trait;
use reqwest::Client;
use serde_json;

use crate::core::{FileService, AppResult, AppError, FileInfo, Pagination};

/// 文件服务实现
/// 
/// 简单的文件服务，支持基本的文件操作
pub struct FileServiceImpl {
    client: Client,
    base_url: String,
}

impl FileServiceImpl {
    /// 创建文件服务
    pub fn new(base_url: &str) -> Self {
        Self {
            client: Client::new(),
            base_url: base_url.to_string(),
        }
    }
    
    /// 构建完整URL
    fn build_url(&self, path: &str) -> String {
        format!("{}{}", self.base_url, path)
    }
    
    /// 添加认证头
    fn add_auth_headers(&self, request: reqwest::RequestBuilder, auth_info: &crate::core::AuthInfo) -> reqwest::RequestBuilder {
        request
            .header("Id", &auth_info.device_id)
            .header("Totp", &auth_info.totp)
    }
}

#[async_trait]
impl FileService for FileServiceImpl {
    async fn list_files(&self, path: &str, pagination: &Pagination) -> AppResult<Vec<FileInfo>> {
        let url = self.build_url("/files");
        
        let response = self.client.get(&url)
            .query(&["path", path])
            .query(&["page", &pagination.page.to_string()])
            .query(&["page_size", &pagination.page_size.to_string()])
            .send()
            .await
            .map_err(|e| AppError::Network(format!("请求失败: {}", e)))?;
        
        if !response.status().is_success() {
            let error_msg = response.text()
                .await
                .unwrap_or_else(|_| "获取文件列表失败".to_string());
            return Err(AppError::FileOperation(error_msg));
        }
        
        let files: Vec<FileInfo> = response.json()
            .await
            .map_err(|e| AppError::Network(format!("解析响应失败: {}", e)))?;
        
        Ok(files)
    }
    
    async fn create_directory(&self, path: &str) -> AppResult<()> {
        let url = self.build_url("/files/directories");
        
        let payload = serde_json::json!({"path": path});
        
        let response = self.client.post(&url)
            .json(&payload)
            .send()
            .await
            .map_err(|e| AppError::Network(format!("请求失败: {}", e)))?;
        
        if !response.status().is_success() {
            let error_msg = response.text()
                .await
                .unwrap_or_else(|_| "创建目录失败".to_string());
            return Err(AppError::FileOperation(error_msg));
        }
        
        Ok(())
    }
    
    async fn delete(&self, path: &str) -> AppResult<()> {
        let url = self.build_url(&format!("/files/{}", urlencoding::encode(path)));
        
        let response = self.client.delete(&url)
            .send()
            .await
            .map_err(|e| AppError::Network(format!("请求失败: {}", e)))?;
        
        if !response.status().is_success() {
            let error_msg = response.text()
                .await
                .unwrap_or_else(|_| "删除失败".to_string());
            return Err(AppError::FileOperation(error_msg));
        }
        
        Ok(())
    }
    
    async fn rename(&self, old_path: &str, new_path: &str) -> AppResult<()> {
        let url = self.build_url("/files/rename");
        
        let payload = serde_json::json!({
            "old_path": old_path,
            "new_path": new_path
        });
        
        let response = self.client.post(&url)
            .json(&payload)
            .send()
            .await
            .map_err(|e| AppError::Network(format!("请求失败: {}", e)))?;
        
        if !response.status().is_success() {
            let error_msg = response.text()
                .await
                .unwrap_or_else(|_| "重命名失败".to_string());
            return Err(AppError::FileOperation(error_msg));
        }
        
        Ok(())
    }
    
    async fn move_item(&self, from_path: &str, to_path: &str) -> AppResult<()> {
        let url = self.build_url("/files/move");
        
        let payload = serde_json::json!({
            "from_path": from_path,
            "to_path": to_path
        });
        
        let response = self.client.post(&url)
            .json(&payload)
            .send()
            .await
            .map_err(|e| AppError::Network(format!("请求失败: {}", e)))?;
        
        if !response.status().is_success() {
            let error_msg = response.text()
                .await
                .unwrap_or_else(|_| "移动失败".to_string());
            return Err(AppError::FileOperation(error_msg));
        }
        
        Ok(())
    }
    
    async fn copy(&self, from_path: &str, to_path: &str) -> AppResult<()> {
        let url = self.build_url("/files/copy");
        
        let payload = serde_json::json!({
            "from_path": from_path,
            "to_path": to_path
        });
        
        let response = self.client.post(&url)
            .json(&payload)
            .send()
            .await
            .map_err(|e| AppError::Network(format!("请求失败: {}", e)))?;
        
        if !response.status().is_success() {
            let error_msg = response.text()
                .await
                .unwrap_or_else(|_| "复制失败".to_string());
            return Err(AppError::FileOperation(error_msg));
        }
        
        Ok(())
    }
    
    async fn download(&self, path: &str) -> AppResult<Vec<u8>> {
        let url = self.build_url(&format!("/files/download/{}", urlencoding::encode(path)));
        
        let response = self.client.get(&url)
            .send()
            .await
            .map_err(|e| AppError::Network(format!("请求失败: {}", e)))?;
        
        if !response.status().is_success() {
            let error_msg = response.text()
                .await
                .unwrap_or_else(|_| "下载失败".to_string());
            return Err(AppError::FileOperation(error_msg));
        }
        
        let data = response.bytes()
            .await
            .map_err(|e| AppError::Network(format!("读取数据失败: {}", e)))?
            .to_vec();
        
        Ok(data)
    }
    
    async fn upload(&self, path: &str, data: &[u8]) -> AppResult<()> {
        let url = self.build_url(&format!("/files/upload/{}", urlencoding::encode(path)));
        
        let response = self.client.post(&url)
            .body(data.to_vec())
            .send()
            .await
            .map_err(|e| AppError::Network(format!("请求失败: {}", e)))?;
        
        if !response.status().is_success() {
            let error_msg = response.text()
                .await
                .unwrap_or_else(|_| "上传失败".to_string());
            return Err(AppError::FileOperation(error_msg));
        }
        
        Ok(())
    }
}